### PR TITLE
cEP-0005: Definition of aspects and tastes

### PR DESCRIPTION
**Pull request by techievena**
Definition of how aspects and tastes should be named.

Closes https://github.com/coala/documentation/issues/223

### CODE REVIEW

**Comment by techievena** [comment]
@coala/aspects-developers Please review

File: cEP-0005.md [hunk 1]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
...
 ```
 
+Before moving any further, it is quite necessary to have a clear-cut definition of both aspects and tastes, to avoid any kind of discrepancies and chaos in the future.
+
+## aspects
+An aspect should just talk about a property of something, without qualifying it. It should be named something like `Length` rather than `TooLong`.
+Then it should be given the tastes. For e.g. aspect `Metadata.CommitMessage.Shortlog.Length` can have a taste like `max_shortlog_length`.
+
+## tastes
+tastes are the values that are meant to change the behaviour of their corresponding aspects. It should not be something like `allow` or `check` which turns the aspect value on or off, but it should be something more like a value which changes the behaviour of an aspect. It can be given as much context as needed but not more. For e.g. for the aspect `Metadata.CommitMessage.Shortlog`, set the taste `shortlog_colon` as true if it must have a colon in the shortlog, and false if the opposite is true.
```
Comment (adhikasp) [inline cEP-0005.md]: > It should not be something like `allow` or `check` which turns the aspect value on or off, but it should be something more like a value which changes the behaviour of an aspect.

You just repeating here. Maybe explain that "taste should describe an aspect in a specific, measurable way"?

File: cEP-0005.md [hunk 2]
```diff
@@ -49,9 +49,9 @@ paragraph:
   not a program and must remain flow insensitive.
 - To specify, what analysis should be run the user will not have to care about
   bears. Instead, an `aspects` setting will be provided allowing the user to
-  specify what aspects of their code should be analyzed. Aspects could be
+  specify what aspects of their code should be analyzed. aspects could be
   `spelling` or `smells` and are namespaced such that `redundancy.clones` will
```
Comment (userzimmermann) [inline cEP-0005.md]: aspect names should also be in `CamelCase` like their class names here. To stay consistent throughout the document...

File: cEP-0005.md [hunk 3]
```diff
@@ -104,7 +104,7 @@ class RedundancyBear(LocalBear):
     # ... and detect code clones.
     DETECT_ASPECTS = {Root.Redundancy.Clone}
```
Comment (userzimmermann) [inline cEP-0005.md]: You could also change this to our new meta style as part of this PR

File: cEP-0005.md [hunk 4]
```diff
@@ -104,7 +104,7 @@ class RedundancyBear(LocalBear):
...
 
-    # Aspects are passed as parameter
+    # aspects are passed as parameter
```
Comment (userzimmermann) [inline cEP-0005.md]: You should somehow make clear that we pass aspect instances here in contrast to the aspect classes that are passed to the bear class definition

File: cEP-0005.md [hunk 5]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
...
+
+## aspects
+An aspect should just talk about a property of something, without qualifying it. It should be named something like `Length` rather than `TooLong`.
+Then it should be given the tastes. For e.g. aspect `Metadata.CommitMessage.Shortlog.Length` can have a taste like `max_shortlog_length`.
+
+## tastes
+tastes are the values that are meant to change the behaviour of their corresponding aspects. It should not be something like `allow` or `check` which turns the aspect value on or off, but it should be something more like a value which changes the behaviour of an aspect. It can be given as much context as needed but not more. For e.g. for the aspect `Metadata.CommitMessage.Shortlog`, set the taste `shortlog_colon` as true if it must have a colon in the shortlog, and false if the opposite is true.
+
+To make things more clear, we can draw an analogy with the real world. For example, if we consider a Tall Man; the aspect of consideration is Length and the corresponding taste value is Tall.
+
```
Comment (userzimmermann) [inline cEP-0005.md]: You should align your line lengths to the rest of the text

**Review by userzimmermann — CHANGES_REQUESTED** [review]
@Techievena Thanks for those changes and additions! We are getting closer to our final manifesto ;) But I would change some little things in this PR...

**Comment by userzimmermann** [comment]
This also supersedes https://github.com/coala/cEPs/pull/67

@EverWinter23 Are you fine with closing your PR in favor of this?

@Techievena Or you could rebase your commit onto @EverWinter23's so that it was not worthless ;)

File: cEP-0005.md [hunk 6]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
             yield Result.from_values(aspect=aspect, ...)
 ```
 
+Before moving any further, it is quite necessary to have a clear-cut definition of both aspects and tastes, to avoid any kind of discrepancies and chaos in the future.
+
+## aspects
```
Comment (sils) [inline cEP-0005.md]: newline after heading please

File: cEP-0005.md [hunk 7]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
             yield Result.from_values(aspect=aspect, ...)
 ```
 
+Before moving any further, it is quite necessary to have a clear-cut definition of both aspects and tastes, to avoid any kind of discrepancies and chaos in the future.
```
Comment (sils) [inline cEP-0005.md]: I'm for a more concrete approach, this whole sentence doesn't add any info, the title in line 129 could hold this info easily. `## Naming of aspects and tastes`

Comment (sils) [reply cEP-0005.md]: i.e. less text, more info!

File: cEP-0005.md (hunk 1 above)
Comment (sils) [inline cEP-0005.md]: >  meant to change the behaviour of their corresponding aspects

aspects are aspects and don't mirror behaviour right?

Comment (sils) [inline cEP-0005.md]: `Metadata.CommitMessage.Shortlog` would have lots of subaspects, the colon would be just one of the subaspects which would need a taste so I feel like this example is actually not good

File: cEP-0005.md [hunk 8]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
...
+
+## aspects
+An aspect should just talk about a property of something, without qualifying it. It should be named something like `Length` rather than `TooLong`.
+Then it should be given the tastes. For e.g. aspect `Metadata.CommitMessage.Shortlog.Length` can have a taste like `max_shortlog_length`.
+
+## tastes
+tastes are the values that are meant to change the behaviour of their corresponding aspects. It should not be something like `allow` or `check` which turns the aspect value on or off, but it should be something more like a value which changes the behaviour of an aspect. It can be given as much context as needed but not more. For e.g. for the aspect `Metadata.CommitMessage.Shortlog`, set the taste `shortlog_colon` as true if it must have a colon in the shortlog, and false if the opposite is true.
+
+To make things more clear, we can draw an analogy with the real world. For example, if we consider a Tall Man; the aspect of consideration is Length and the corresponding taste value is Tall.
```
Comment (sils) [inline cEP-0005.md]: this isn't really helping me personally gather a clearer understanding of aspects and their naming

File: cEP-0005.md (hunk 1 above)
Comment (adtac) [reply cEP-0005.md]: Also, I'm not entirely sure why we can't have a taste called `allow`. An aspect `Metadata.CommitMessage.Shortlog.Colon` will need a boolean taste called `allow`.

**Comment by sils** [comment]
Because allow accepts both a colon and not having a colon. If the taste is
active it should either enforce a colon or enforce no colon and not "allow"
both, in the allow case the aspect is useless and schould not be selected

File: cEP-0005.md (hunk 1 above)
Comment (userzimmermann) [reply cEP-0005.md]: @adtac To make it more verbose and avoid taste naming clashes, so that you can for example request all suaspects of `Shortlog` by just defining to check `Metadata.CommitMessage.Shortlog`, or even the whole `Metadata.CommitMessage` subtree, and then just set a value for `shortlog_colon` without having to explicitly specify the `Shortlog.ColonExistence` aspect

@Techievena You should also add more information about that stuff

**CC** @sils

File: cEP-0005.md [hunk 9]
```diff
@@ -124,6 +124,17 @@ class RedundancyBear(LocalBear):
             yield Result.from_values(aspect=aspect, ...)
 ```
 
+Before moving any further, it is quite necessary to have a clear-cut definition of both aspects and tastes, to avoid any kind of discrepancies and chaos in the future.
+
+## aspects
+An aspect should just talk about a property of something, without qualifying it. It should be named something like `Length` rather than `TooLong`.
+Then it should be given the tastes. For e.g. aspect `Metadata.CommitMessage.Shortlog.Length` can have a taste like `max_shortlog_length`.
+
+## tastes
```
Comment (userzimmermann) [inline cEP-0005.md]: Same as @sils mentioned for `##aspects` heading in https://github.com/coala/cEPs/pull/72#discussion_r113806973

**Review by userzimmermann — CHANGES_REQUESTED** [review]
One tiny little more thing :)

**Comment by techievena** [comment]
@userzimmermann Regarding Currently I'm modifying this PR only not to lose the discussions. When everything is finalized, I'll rebase with https://github.com/coala/cEPs/pull/67. :smiley:

File: cEP-0005.md (hunk 7 above)
Comment (techievena) [reply cEP-0005.md]: @sils My first time with docs. Sorry if I'm screwing up at this.

File: cEP-0005.md (hunk 8 above)
Comment (techievena) [reply cEP-0005.md]: So should I just remove this part or come up with a better example or just rephrase this example only to make it understandable?

File: cEP-0005.md (hunk 1 above)
Comment (techievena) [reply cEP-0005.md]: @adtac We cannot have a taste value like `allow` bcoz using a taste like `allow` it sounds more like we are enabling or disabling an aspect rather than setting it value to True or False and taste values should never control if an aspect is enabled or disabled. This is the only reason we must have a proper naming for aspects and tastes such that they don't deviate from their real purpose and it can really be chaotic in future if they do so.

File: cEP-0005.md (hunk 2 above)
Comment (techievena) [reply cEP-0005.md]: @userzimmermann Will fix these in the next iteration :smile:

File: cEP-0005.md [hunk 10]
```diff
@@ -124,6 +124,32 @@ class RedundancyBear(LocalBear):
             yield Result.from_values(aspect=aspect, ...)
 ```
 
+## Naming of aspects and tastes
+
+## aspects
```
Comment (userzimmermann) [inline cEP-0005.md]: Better `### aspects`

File: cEP-0005.md [hunk 11]
```diff
@@ -124,6 +124,32 @@ class RedundancyBear(LocalBear):
...
+
+An aspect should just talk about a property of something, without qualifying
+it. It should be named something like `Length` rather than `TooLong`. Then it
+should be given the tastes.
+For e.g. an aspect can be named as `Metadata.CommitMessage.Shortlog.Length` to
+represent the length property of the shortlog (first line) of a commit message.
+
+## tastes
```
Comment (userzimmermann) [inline cEP-0005.md]: Better `### tastes`

File: cEP-0005.md [hunk 12]
```diff
@@ -124,6 +124,32 @@ class RedundancyBear(LocalBear):
...
+For e.g. for the aspect `Metadata.CommitMessage.Shortlog.`
+`FirstCharacter`, set the taste `shortlog_starts_upper_case` as true if the
+shortlog must begin with a upper case letter consistently, and false if the
+opposite is true.
+
+__Note:__ aspects and tastes are orthogonal concepts and should never overlap.
+Therefore extreme caution must be taken while naming aspects and tastes to
+avoid any discrepancy and inconsistency in the future.
+
 The concept of aspects allows us to implement a consistency check and reduce
```
Comment (userzimmermann) [inline cEP-0005.md]: A new section starts here, not related to naming anymore. So we need some `##` heading at this point

File: cEP-0005.md (hunk 3 above)
Comment (userzimmermann) [reply cEP-0005.md]: @Techievena You should really change this to our new meta `bearclass` style

And you should also change it to use new `languages=` keyword in `class` definition args according to https://github.com/coala/coala/issues/4111

**Review by userzimmermann — CHANGES_REQUESTED** [review]
@Techievena Sorry ;) but there some other things left... little things of course :)

Thanks for the effort you put into this!

File: cEP-0005.md [hunk 13]
```diff
@@ -96,15 +96,13 @@ from coalib.results.Result import Result
...
-    LANGUAGES = {"C", "Python"}
-    # This bear is supposed to be able to fix unused imports...
-    FIX_ASPECTS = {Root.Redundancy.UnusedImport}
-    # ... and detect code clones.
-    DETECT_ASPECTS = {Root.Redundancy.Clone}
+# To define the aspects and languages for a bear using the bear metaclass
+class RedundancyBear(LocalBear, aspects={
+    'detect': [Root.Redundancy.Clone],
+    'fix': [Root.Redundancy.UnusedImport]},
+    languages: ['C', 'Python']):
```
Comment (userzimmermann) [inline cEP-0005.md]: Little syntax error ;) Must be `languages=['C', 'Python']`. And IMO, it would overall look better and PEP8ier like:

```python
class RedundancyBear(LocalBear, aspects={
        'detect': [Root.Redundancy.Clone],
        'fix': [Root.Redundancy.UnusedImport],
}, languages=['C', 'Python']):
    ...
```

**Review by userzimmermann — CHANGES_REQUESTED** [review]
@Techievena Last request! Promise ;P

File: cEP-0005.md (hunk 13 above)
Comment (techievena) [reply cEP-0005.md]: Oops my bad!!! :worried:

File: cEP-0005.md [hunk 14]
```diff
@@ -96,15 +96,13 @@ from coalib.results.Result import Result
...
-    LANGUAGES = {"C", "Python"}
-    # This bear is supposed to be able to fix unused imports...
-    FIX_ASPECTS = {Root.Redundancy.UnusedImport}
-    # ... and detect code clones.
-    DETECT_ASPECTS = {Root.Redundancy.Clone}
+# To define the aspects and languages for a bear using the bear metaclass
+class RedundancyBear(LocalBear, aspects={
+    'detect': [Root.Redundancy.Clone],
+    'fix': [Root.Redundancy.UnusedImport]
+    },languages=['C', 'Python']):
```
Comment (userzimmermann) [inline cEP-0005.md]: Still not fully PEP8 compliant :/ Patch from `PEP8Bear`:

```diff
 class RedundancyBear(LocalBear, aspects={
-    'detect': [Root.Redundancy.Clone],
-    'fix': [Root.Redundancy.UnusedImport]
-    },languages=['C', 'Python']):
+        'detect': [Root.Redundancy.Clone],
+        'fix': [Root.Redundancy.UnusedImport]
+        }, languages=['C', 'Python']):
```

In statement expressions creating a new scope below you should always use 8 spaces for continuation... To not make it look like the actual scope below, which is indented by 4

You can also move the closing `aspects={...}` brace to the beginning of the line:

```python
class ...(..., aspects={
        ...
}, languages=['C', 'Python']):
```

But that's just a matter of taste ;)

**Review by userzimmermann — CHANGES_REQUESTED** [review]
@Techievena Code example still not fully PEP8 compliant :/ But that's now really the last complaint! :)

**Review by userzimmermann — APPROVED** [review]

**Comment by userzimmermann** [comment]
@Techievena Perfect :) Only one question left: Still wanna do rebasing according to https://github.com/coala/cEPs/pull/72#issuecomment-298012251 ? Otherwise we just directly merge

**Comment by userzimmermann** [comment]
ack https://github.com/coala/cEPs/pull/72/commits/f6a01705471ad897a280be272a0750caa46e4c54

**Comment by userzimmermann** [comment]
@Techievena Thanks again for all the effort! 🎉

**Comment by userzimmermann** [comment]
@rultor merge

**Comment by meetmangukiya** [comment]
Please rebase @techievena

**Comment by sils** [comment]
cherry pick bot as well? If it works...

File: cEP-0005.md [hunk 15]
```diff
@@ -49,9 +49,9 @@ paragraph:
   not a program and must remain flow insensitive.
 - To specify, what analysis should be run the user will not have to care about
   bears. Instead, an `aspects` setting will be provided allowing the user to
-  specify what aspects of their code should be analyzed. Aspects could be
-  `spelling` or `smells` and are namespaced such that `redundancy.clones` will
-  be possible as well. Aspects can be defined inclusive (analyze those aspects,
+  specify what aspects of their code should be analyzed. aspects could be
```
Comment (jayvdb) [inline cEP-0005.md]: Use `Aspects` at the start of the sentence (or move the word away from the start of the sentence.).

File: cEP-0005.md [hunk 16]
```diff
@@ -225,7 +251,7 @@ This allows to show more information in results. Bears can give a custom result
 message; however, by default the message can be generated from the settings of
 the aspects with the ``result_message`` function.
 
-Aspects can also define a ``style_paragraph`` function that returns a small
```
Comment (jayvdb) [inline cEP-0005.md:241]: undo this change.  It doesnt belong as part of your commit.
I've no objection to a new PR which alters the structure of this sentence such that the word aspects is not at the beginning of the sentence.

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: cEP-0005.md (hunk 15 above)
Comment (userzimmermann) [reply cEP-0005.md]: @jayvdb Please please please let's not restart the discussion about the `aspects` spelling now :) This is counterproductive for GSoC ;) The **#aspectsTeam** has made the decision to keep it always-lower-case... And this is one of those little important fancy (maybe also funny, but anyway... even better) aspects that define the strong bonding inside the team

cc @coala/aspects-developers @sils @meetmangukiya @adtac

Comment (userzimmermann) [reply cEP-0005.md]: @Techievena just put all *aspect(s)* appearances into backticks --> `aspect` / `aspects`. Then they are literals and don't fall under those grammar rules ;)

cc @coala/aspects-developers

File: cEP-0005.md (hunk 16 above)
Comment (userzimmermann) [reply cEP-0005.md:241]: @Techievena Same as https://github.com/coala/cEPs/pull/72/files#r120339847

**Comment by userzimmermann** [comment]
@Techievena Thanks again for that stuff! Those additions are important and also useful for @adhikasp's GSoC work. He can benefit from it... So let's get it finally approved and merged

cc @jayvdb @coala/aspects-developers

File: cEP-0005.md [hunk 17]
```diff
@@ -137,6 +135,34 @@ class RedundancyBear(LocalBear):
...
+For e.g. an aspect can be named as `Metadata.CommitMessage.Shortlog.Length` to
+represent the length property of the shortlog (first line) of a commit message.
+
+#### tastes
+
+tastes are the values that are used to initialize the aspect settings. It
+should not be something like `allow` or `check` which enables or disables an
+aspect, rather it should be something more like a value which describes an
+aspect in a specific, measurable way. It can be given as much context as
+needed but not more.<br/>
```
Comment (sils) [inline cEP-0005.md]: remove html, just put another linebreak instead

File: cEP-0005.md [hunk 18]
```diff
@@ -137,6 +135,34 @@ class RedundancyBear(LocalBear):
...
+For e.g. for the aspect `Metadata.CommitMessage.Shortlog.`
+`FirstCharacter`, set the taste `shortlog_starts_upper_case` as true if the
```
Comment (sils) [inline cEP-0005.md]: FirstCharacter isn't an aspect, FirstCharacterCasing could be. Right @Asnelchristian @userzimmermann ?

**Comment by sils** [comment]
@Techievena this is already pretty good, with the open comments resolved I think we can merge that :+1:

**Comment by jayvdb** [comment]
unack https://github.com/coala/cEPs/pull/72/commits/458cfd57b6727b9d9ee991fda88ec9cf3bfefdf1

**Comment by adhikasp** [comment]
ack 344ae64301daa8a68c02005a5655eca5f1c592aa

**Review by jayvdb — APPROVED** [review]

**Comment by jayvdb** [comment]
ack 344ae64

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot ff
